src/integrationTest/java/com/mongodb/hibernate/EmbeddableIntegrationTests.java

```diff
@@ -0,0 +1,415 @@
+        })
+@ExtendWith(MongoExtension.class)
+class EmbeddableIntegrationTests implements SessionFactoryScopeAware {
+    @InjectMongoCollection("items")
+    private static MongoCollection<BsonDocument> mongoCollection;
+
+    private SessionFactoryScope sessionFactoryScope;
+
+    @Test
+    void testFlattenedValues() {
+        var item = new ItemWithFlattenedValues();
+        {
+            item.id = new EmbeddableValue();
+            item.id.a = 1;
+            item.flattened1 = new EmbeddableValue();
+            item.flattened1.a = 2;
+            item.flattened2 = new EmbeddablePairValue1();
+            item.flattened2.a = 3;
+            item.flattened2.flattened = new EmbeddablePairValue2(4, 5);
+            item.flattened2.parent = item;
+        }
+        sessionFactoryScope.inTransaction(session -> session.persist(item));
+        assertThat(mongoCollection.find())
+                .containsExactly(new BsonDocument()
+                        .append(ID_FIELD_NAME, new BsonInt32(item.id.a))
+                        .append("flattened1_a", new BsonInt32(2))
+                        .append("flattened2_a", new BsonInt32(3))
+                        .append("flattened2_flattened_a", new BsonInt32(4))
+                        .append("flattened2_flattened_b", new BsonInt32(5)));
+        var loadedItem =
+                sessionFactoryScope.fromTransaction(session -> session.find(ItemWithFlattenedValues.class, item.id));
+        assertEquals(item, loadedItem);
+        var updatedItem = sessionFactoryScope.fromTransaction(session -> {
+            var result = session.find(ItemWithFlattenedValues.class, item.id);
+            result.flattened1.a = -result.flattened1.a;
+            return result;
+        });
+        loadedItem =
+                sessionFactoryScope.fromTransaction(session -> session.find(ItemWithFlattenedValues.class, item.id));
+        assertEquals(updatedItem, loadedItem);
+    }
+
+    @Test
+    void testFlattenedEmptyValue() {
+        var item = new ItemWithFattenedEmptyValue();
+        {
+            item.id = 1;
+            item.omitted = new EmbeddableEmptyValue();
+        }
+        sessionFactoryScope.inTransaction(session -> session.persist(item));
+        assertThat(mongoCollection.find())
+                .containsExactly(
+                        // Hibernate ORM does not store/read the empty `item.omitted` value.
+                        // See https://hibernate.atlassian.net/browse/HHH-11936 for more details.
+                        new BsonDocument(ID_FIELD_NAME, new BsonInt32(item.id)));
+        var loadedItem =
+                sessionFactoryScope.fromTransaction(session -> session.find(ItemWithFattenedEmptyValue.class, item.id));
+        assertUsingRecursiveComparison(item, loadedItem, (assertion, actual) -> assertion
+                .ignoringFields("omitted")
+                .isEqualTo(actual));
```

Comment:
The `actual` parameter, despite always being the same as `loadedItem`, is needed to work around the effectively `final` requirement for a variable to be allowed to be put in a lambda expression. The Java SE API uses the same approach in some places, for example in `Map.merge`.